dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java

```diff
@@ -47,35 +59,119 @@
+
+                state = beginProbation(endpoint);
+            }
+
+            if (state instanceof Probation) {
+                Probation probation = (Probation) state;
+                if (probation.acquireStartPermit()) {
+                    log.debug("Probation channel request allowed");
+                    return delegate.maybeExecute(endpoint, request).map(future -> DialogueFutures.addDirectCallback(
+                            future, new BlacklistingCallback(Optional.of(probation), endpoint)));
+                } else {
+                    log.debug("Probation channel request not allowed");
+                    return Optional.empty();
+                }
+            }
         }
+
+        return delegate.maybeExecute(endpoint, request).map(future ->
+                DialogueFutures.addDirectCallback(future, new BlacklistingCallback(Optional.empty(), endpoint)));
+    }
+
+    private void blacklist(Endpoint endpoint) {
+        perEndpointBlacklistState.put(endpoint, new BlacklistUntil(ticker.read() + duration.toNanos()));
+    }
+
+    private Probation beginProbation(Endpoint endpoint) {
+        Probation probation = new Probation(NUM_PROBATION_REQUESTS);
+        perEndpointBlacklistState.put(endpoint, probation);
+        return probation;
+    }
+
+    private void probationComplete(Endpoint endpoint) {
+        perEndpointBlacklistState.invalidate(endpoint);
     }
 
     private final class BlacklistingCallback implements FutureCallback<Response> {
+        private final Optional<Probation> probationPermit;
+        private final Endpoint endpoint;
+
+        private BlacklistingCallback(Optional<Probation> probationPermit, Endpoint endpoint) {
+            this.probationPermit = probationPermit;
+            this.endpoint = endpoint;
+        }
+
         @Override
         public void onSuccess(Response response) {
             // TODO(jellis): use the Retry-After header (if present) to determine how long to blacklist the channel
-            if (response.code() == 503) {
-                isBlacklisted.put(KEY, VALUE);
+            if (response.code() == 503 || response.code() == 500) {
+                log.debug(
+                        "Blacklisting {} due to status code {}",
+                        UnsafeArg.of("delegate", delegate),
+                        SafeArg.of("code", response.code()));
+                blacklist(endpoint);
+            } else if (probationPermit.isPresent() && probationPermit.get().checkIfProbationIsComplete()) {
+                log.debug("Probation is complete");
+                probationComplete(endpoint);
             }
         }
 
         @Override
         public void onFailure(Throwable _throwable) {
-            isBlacklisted.put(KEY, VALUE);
+            blacklist(endpoint);
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "BlacklistingChannel{" + delegate + '}';
+    }
+
+    // I wish java had union types
+    interface BlacklistState {}
```

Comment:
derive4j is your friend here

Reply:
So I actually really don't want to have the extra faff and tooling setup of derive4j - can you forgive these two purely implementation-detail classes?